pkg/apis/kubeone/config/migrate.go

```diff
@@ -17,45 +17,72 @@ limitations under the License.
 	"k8c.io/kubeone/pkg/yamled"
+
+	kyaml "sigs.k8s.io/yaml"
 )
 
-// MigrateOldConfig migrates KubeOneCluster v1beta1 object to v1beta2
-func MigrateOldConfig(clusterFilePath string) (interface{}, error) {
-	oldConfig, err := loadClusterConfig(clusterFilePath)
+// MigrateV1beta2V1beta3 migrates KubeOneCluster v1beta2 object to v1beta3
+func MigrateV1beta2V1beta3(clusterFilePath string) (*kubeonev1beta3.KubeOneCluster, error) {
+	doc, err := loadClusterConfig(clusterFilePath)
 	if err != nil {
 		return nil, fail.Runtime(err, "loading cluster config to migrate")
 	}
 
 	// Check is kubeone.k8c.io/v1beta2 config provided
-	apiVersion, apiVersionExists := oldConfig.GetString(yamled.Path{"apiVersion"})
+	oldAPIVersion, apiVersionExists := doc.GetString(yamled.Path{"apiVersion"})
 	if !apiVersionExists {
 		return nil, fail.Config(fmt.Errorf("apiVersion not present in the manifest"), "checking apiVersion presence")
 	}
 
-	if apiVersion != kubeonev1beta2.SchemeGroupVersion.String() {
-		return nil, fail.Config(fmt.Errorf("migration is available only for %q API, but %q is given", kubeonev1beta2.SchemeGroupVersion.String(), apiVersion), "checking apiVersion compatibility")
+	if oldAPIVersion != kubeonev1beta2.SchemeGroupVersion.String() {
+		return nil, fail.Config(fmt.Errorf("migration is available only for %q API, but %q is given", kubeonev1beta2.SchemeGroupVersion.String(), oldAPIVersion), "checking apiVersion compatibility")
 	}
 
 	// Ensure kind is KubeOneCluster
-	kind, kindExists := oldConfig.GetString(yamled.Path{"kind"})
+	kind, kindExists := doc.GetString(yamled.Path{"kind"})
 	if !kindExists {
 		return nil, fail.ConfigValidation(fmt.Errorf("kind not present in the manifest"))
 	}
 	if kind != KubeOneClusterKind {
 		return nil, fail.ConfigValidation(fmt.Errorf("migration is available only for kind %q, but %q is given", KubeOneClusterKind, kind))
 	}
 
-	// The APIVersion has been changed to kubeone.k8c.io/v1beta3
-	oldConfig.Set(yamled.Path{"apiVersion"}, kubeonev1beta3.SchemeGroupVersion.String())
+	var (
+		buffer           bytes.Buffer
+		oldManifest      = kubeonev1beta2.NewKubeOneCluster()
+		newManifest      = kubeonev1beta3.NewKubeOneCluster()
+		internalManifest = new(kubeoneapi.KubeOneCluster)
+	)
+
+	if err = yaml.NewEncoder(&buffer).Encode(doc.Root()); err != nil {
+		return nil, fail.Config(err, "marshaling v1beta2 KubeOneCluster")
+	}
+
+	if err = kyaml.UnmarshalStrict(buffer.Bytes(), oldManifest); err != nil {
+		return nil, fail.Runtime(err, "testing unmarshal v1beta2 KubeOneCluster")
+	}
+
+	if err = scheme.Scheme.Convert(oldManifest, internalManifest, kubeoneapi.SchemeGroupVersion); err != nil {
+		return nil, fail.Config(err, "converting v1beta2/KubeOneCluster into internal KubeOneCluster")
+	}
+
+	if err = scheme.Scheme.Convert(internalManifest, newManifest, kubeonev1beta3.SchemeGroupVersion); err != nil {
+		return nil, fail.Config(err, "converting internal KubeOneCluster into v1beta3/KubeOneCluster")
+	}
```

Comment:
I strongly believe we should use `yamled` here instead. There are many problems with this implementation, the first one being the fact that comments are not preserved, the second one that the file will be alphabetically sorted (instead of leaving sort as it is). This is not acceptable for a user-facing command, I'm strong -1 on this change.

Reply:
yamled does not preserve comment ether

Reply:
I see, but at least we get to preserve the key order and the key set, so I believe we still have huge benefits. Thanks for addressing this issue!